File: projects/KD1OUR/automata/automata-scala/src/statemachine/scala/StateMachine.scala
package statemachine.scala

import statemachine.scala.model.StateMachineState._
import statemachine.scala.model.TerminalSymbol._
import statemachine.scala.model.ProductionRules._

/**
 * Az automata algoritmusának implementációja.
 */
object StateMachine {

  def process(input : String) : String = {

    def processSymbols(input : Seq[Option[TerminalSymbol]], state : StateMachineState) : Option[StateMachineState] = {
      // végeztünk, adjuk vissza az utolsó állapotot
      if (input.isEmpty) Some(state)
      else {
        // érvénytelen karakter volt a stringben (nincs hozzá definiált terminális szimbólum)
        if (input contains None) None
        else {
          // az állapotnak és szimbólumnak megfelelően alkalmazzuk a szabályt
          // ellenőrizzük, hogy az adott állapotban az adott szimbólumhoz tartozik-e állapot-átmenet
          val stepResult = applyRule(state)(input.head.get)
          if (!stepResult.newState.isDefined) None
          else {
			// nézzük meg, hogy az adott állapot-átmenet feldolgozza-e a karaktert, vagy láncszabály
            if(stepResult.terminalProcessed) processSymbols(input.tail, stepResult.newState.get)
            else processSymbols(input, stepResult.newState.get)
          }
        }
      }
    }

    // input ketté bontása (e, vagy E karakter előtti rész, és a többi)
    val indexOfE = Integer.max(input.indexOf('e'), input.indexOf('E'))
    val inputUntilE = if (indexOfE > -1) input.substring(0, indexOfE) else input : String
    val outputFromE = if (indexOfE > -1) input.substring(indexOfE) else "" : String

    // a karaktereket megfeleltetjük terminális szimbúloknak, majd feldolgozzuk őket a fent definiált process függvénnyel
    val terminalSymbolsUntilE = inputUntilE.map(x => getTerminalSymbol(x))
    val terminalSymbolsFromE = outputFromE.map(x => getTerminalSymbol(x))
    // elmentjük az E karakter előtt lévő állapotot, és az utolsó állapotot
    val stateBeforeE = processSymbols(terminalSymbolsUntilE, START_SIGNED)
    val lastState = if (stateBeforeE.isDefined) processSymbols(terminalSymbolsFromE, stateBeforeE.get) else None

    // amennyiben valamely karakter érvénytelen volt (vagy mert nem egy érvényes terminális, vagy mert olyan állapotban
	// következett, ahol nem következhet), vagy az utolsó állapot nem elfogadó állapot, akkor az eredmény "FAIL"
    if(!lastState.isDefined || !isFinalState(lastState.get)) "FAIL" else {
      // egy kis output formázás, ha az automata elfogadta az inputot
	    // előjel levágása
      val outputUntilE = if(terminalSymbolsUntilE.head.get == SIGN) inputUntilE.tail else inputUntilE

	  // hiányzó 0-k pótlása a szám elejéről/végéről
	  // a korrigálás mikéntje kitalálható az utolsó állapot és terminális szimbólum ismeretében
      stateBeforeE.get match {
        case LEGAL_STATE_1 => "OK 0" + outputUntilE + outputFromE
        case LEGAL_STATE_2 => "OK 0" + outputFromE
        case LEGAL_STATE_3 => if (terminalSymbolsUntilE.last.get != DIGIT) "OK 0" else "OK " + outputUntilE + outputFromE
        case LEGAL_STATE_4 => "OK " + outputUntilE + ".0" + outputFromE
        case LEGAL_STATE_5 => "OK " + outputUntilE + (if (terminalSymbolsUntilE.last.get != DIGIT) "0" else "") + outputFromE
        case _ => "FAIL"
      }
    }
  }
}

File: projects/KD1OUR/automata/automata-java/src/main/java/statemachine/java/StateMachine.java
package statemachine.java;

import statemachine.java.model.StateMachineState;
import statemachine.java.model.TerminalSymbol;

import java.util.Map;

/**
 * Az automata működését magába foglaló osztály.
 */
public class StateMachine {

    private StateMachineState state;
    private final StateMachineState start;
    private final Map<StateMachineState, ProductionRuleRHS> productionRules;

    public StateMachine(StateMachineState start, Map<StateMachineState, ProductionRuleRHS> productionRules){
        this.start = start;
        this.productionRules = productionRules;
    }

    public String parse(String input){
        StringBuilder sb = new StringBuilder();

        char currentChar;
        TerminalSymbol currentTerminal = null;
        ProductionRuleRHS currentRule;
        StateMachineState nextState;

        this.state = this.start;

        for(int i = 0; i < input.length(); i++){
            currentChar = input.charAt(i);
            currentTerminal = TerminalSymbol.valueOf(currentChar);

            // érvénytelen terminális esetén FAIL
            if(currentTerminal == null){
                return "FAIL";
            }

            currentRule = this.productionRules.get(this.state);
            nextState = currentRule.getNewState(currentTerminal);

            // ha a terminálist nem fogadja el a szabály
            if(nextState == null){
                // megnézzük van-e láncszabály
                nextState = currentRule.getChainRuleNextState();
                if(nextState == null){
                    // ha ez sincs, akkor érvénytelen input
                    return "FAIL";
                } else {
                    // egyébként marad az eddigi input és állapotot váltunk
                    this.state = nextState;
                    i--;
                    continue;
                }
            }

            if (nextState == StateMachineState.ILLEGAL_STATE_0 && sb.length() == 0) {
                sb.append('0');
            }

            // egyébként (ha találtunk érvényes állapot-átmenetet):
            // eltároljuk a karaktert (hozzá fog tartozni az eredményhez, ha nem előjel), és állapotot váltunk
            if (currentTerminal == TerminalSymbol.EXPONENTIAL) {
                sb.append('e');
            } else if(currentChar == '-' && nextState == StateMachineState.LEGAL_STATE_6) {
                sb.append(currentChar);
            } else if(currentTerminal != TerminalSymbol.SIGN) {
                sb.append(currentChar);
            }


            this.state = nextState;
        }

        // végeztünk az input feldolgozásával, megnézzük, hogy elfogadó állapotba jutottunk-e
        if(this.state.isFinalState()){
            // a befejező állapottól függően még picit alakítani kell az eredményen
            switch(this.state){
                case LEGAL_STATE_1:
                    return "OK " + sb.toString();
                case LEGAL_STATE_2:
                    return "OK 0";
                case LEGAL_STATE_4:
                    return "OK " + sb.toString() + ".0";
                case LEGAL_STATE_3:
                    return (currentTerminal != TerminalSymbol.DIGIT) ? "OK 0" : ("OK " + sb.toString());
                case LEGAL_STATE_5:
                    return "OK " + sb.toString() + (currentTerminal != TerminalSymbol.DIGIT && currentTerminal != TerminalSymbol.ZERO ? "0" : "");
                case LEGAL_STATE_6:
                    return "OK " + sb.toString();
            }
        }

        return "FAIL";
    }
}
